File: starter_app/app/api/builds.py
from flask_login import LoginManager, current_user, login_user, logout_user, login_required  # noqa
from flask import (Blueprint, jsonify, url_for, request, redirect, render_template)  # noqa
from app.models import Build, db, User
from sqlalchemy import or_

builds = Blueprint('builds', __name__)


@builds.route("/<userId>")
def get_builds(userId):
    builds = db.session.query(Build).filter(Build.user_id == 1)
    format_builds = {"builds": {}, "userbuilds": {}}
    if(userId != '1'):
        userbuilds = db.session.query(Build).filter(Build.user_id == userId)
        for userbuild in userbuilds:
            format_builds["userbuilds"][userbuild.id] = userbuild.to_dict()
    for build in builds:
        format_builds["builds"][build.id] = build.to_dict()
    return format_builds


@builds.route("/", methods=["POST"])
def create_build():
    data = request.json
    new_build = Build(
        title=data['title'],
        description=data['description'],
        user_id=data['userId'],
        caseId=data['caseId'],
        motherBoardId=data['motherBoardId'],
        cpuId=data['cpuId'],
        coolerId=data['coolerId'],
        hardDriveId=data['hardDriveId'],
        ramId=data['ramId'],
        gpuId=data['gpuId'],
        powerSupplyId=data['powerSupplyId'],
        networkCardId=data['networkCardId'],
    )
    db.session.add(new_build)
    db.session.commit()
    format_build = new_build.to_dict()
    return {"build": format_build}


@builds.route("/delete/<buildId>", methods=["DELETE"])
def delete_build(buildId):
    build = db.session.query(Build).filter(Build.id == buildId).first()
    db.session.delete(build)
    db.session.commit()

File: client/src/store/builds.js
import Cookies from 'js-cookie'

const GET_BUILDS = 'auth/GET_BUILDS';
const DELETE_BUILD = 'auth/DELETE_BUILD';
const CREATE_BUILD = 'auth/CREATE_BUILD';

const getBuilds = (builds) => {
    return {
        type: GET_BUILDS,
        builds,
    };
};


const createBuild = (build) => ({
    type: CREATE_BUILD,
    build
})

const removeBuild = (buildId) => {
    return {
        type: DELETE_BUILD,
        buildId
    };
};


export const fetchBuilds = (userId) => {
    return async (dispatch) => {
        const res = await fetch(`/api/builds/${userId}`);
        const data = await res.json();
        dispatch(getBuilds(data));
        return res;
    };
};

export const makeBuild = ( title, description, user_id, caseId, motherBoardId, cpuId, coolerId, hardDriveId, ramId, gpuId, powerSupplyId, networkCardId) => {
    return async (dispatch) => {
        const csrfToken = Cookies.get("XSRF-TOKEN");
        const res = await fetch(`/api/builds`, {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
                "X-CSRF-TOKEN": csrfToken,
            },
            body: JSON.stringify({ title, description, user_id, caseId, motherBoardId, cpuId, coolerId, hardDriveId, ramId, gpuId, powerSupplyId, networkCardId })
        });
        const data = await res.json();
        if (res.ok) {
            dispatch(createBuild(data));
            res.data = data;
        }
        return res;
    };
}

export const deleteBuild = (buildId) => async dispatch => {
    const res = await fetch(`/api/builds/delete/${buildId}`, {
        method: "DELETE",
        headers: {
            "Content-Type": "application/json",
        },
    });

    if (res.ok) {
        dispatch(removeBuild(buildId));
    }

    return res;
}


export default function buildsReducer(state = {}, action) {
    Object.freeze(state);
    switch (action.type) {
        case GET_BUILDS:
            return action.builds;
        case CREATE_BUILD:
            return action.build;
        case DELETE_BUILD:
            return action.build;
        default:
            return state;
    }
};
